### PR TITLE
Backend Courses

### PR DESCRIPTION
**Pull request by sjdefran1**
---
#12 - Course Model & Repo

Created the course model and course repository, Course.Java & CourseRepository.java

Added following attributes to Course schema for now

```
views: int
thumbsUp: int
thumbsDown: int
```

**One note/question** I have is with timestamps in SQL and Java, from a few minutes of internet research and testing this seems to work

```java
import org.hibernate.annotations.CreationTimestamp;
import java.sql.Timestamp;

@CreationTimestamp
private Timestamp createdAt;
```

This way these are not passed in as values from client side,  but rather automatically created when the object is created and saved to the table

```
EX timestamp: 2024-02-09 15:58:01.963693
```

---

#13 - Course Controller

The controller has all basic crud operations as well as a few extras.  One thing to notice is I used RequestBody as well which allows for sending in a json object in the body of the request, java then creates an object out of it, in this case a Course object. It allows for easily using getter and  setter methods on the objects as well.

```java
 @PostMapping("/course")
    public ResponseEntity<Course> createCourse(@RequestBody Course course){
        courseRepository.save(course);
        return new ResponseEntity<>(course, HttpStatus.OK);
    }
```
Example request body
```Json
{
    "title": "Testing Course after new changes",
    "description": "da course",
    "category": "Testing",
    "views": 0,
    "thumbsUp": 0,
    "thumbsDown": 0
}
```

-- It is important to note not to send automatically generated values to these endpoints though when creating them, for example when creating a new Course, you won't provide an ID, it will be generated on object creation.

**Some custom functions**

- getAllCourses() - self explantory
- updateCourseView() -- pass in an id and it will automatically update course view count and save, not rlly sure if this is the best way of going at it. But seems annoying to have to update entire object each time we want to increment a view. But this seems like a similar issue with thumbs up and thumbs down. So not sure what to do there.

### CODE REVIEW

File: server/src/main/java/com/skillmagnet/Course/Course.java [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+    private String description;
+    private String category;
+    private int views;
+    private int thumbsUp;
+    private int thumbsDown;
+
+    @CreationTimestamp
+    private Timestamp createdAt;
+    @CreationTimestamp
+    private Timestamp status;
```
Comment (nickerick) [inline server/src/main/java/com/skillmagnet/Course/Course.java]: The `@CreationTimestamp` is fine with me - there's tons of crazy annotations that do fun stuff. Nice find lol.

:question: Is there any reason you did `Timestamp` vs a `DateTime`? I looked it up and I'm not oppposed to the Timestamp as I don't fully understand the differences. Just I've always seen a DateTime used lol. If you can still format it into a date (e.g. Jan 10th, 2024 8:05:00) then I'm good with it.

File: server/src/main/java/com/skillmagnet/Course/CourseController.java [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+     */
+    @Autowired
+    CourseRepository courseRepository;
+
+    /**
+     * Retrieves specific course by id
+     * @param id - id of course to find
+     * @return - course object || NOT_FOUND
+     */
+    @GetMapping("/course")
```
Comment (nickerick) [inline server/src/main/java/com/skillmagnet/Course/CourseController.java]: Hate to sound like a nitpicker here lol but definitely I think is beneficial to have in mind

But generally with REST API styles, you'd want to use a path variable to identity resources and request/query params to do stuff like filtering, paging, etc. Here's an [example](https://www.baeldung.com/spring-pathvariable) of what I mean.

So to get a specific course, it's like hey I want courses, specifically with ID 7 so `GET /course/7`
But if it's hey I just want courses, then `GET /course`

But then the extras (so paging, filtering) are tacked onto the request. So get me all courses (the resource) where the name contains 'Java'. `GET /course?filter=java`

File: server/src/main/java/com/skillmagnet/Course/CourseController.java [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+        Course course = courseRepository.findById(id);
+        if(course == null){
+            return new ResponseEntity<>(null, HttpStatus.NOT_FOUND);
+        }
+        
+        // increment view, save, and return updated info
+        course.setViews(course.getViews() + 1);
+        courseRepository.save(course);
+        return new ResponseEntity<>(course, HttpStatus.OK);
+    }
```
Comment (nickerick) [inline server/src/main/java/com/skillmagnet/Course/CourseController.java:125]: I agree with your comment about updating view/thumbsup/thumbsdown. I think this is a good first pass and good idea to pull these out. 

We can wait to see how FE does the client and adjusts based on that.

**Review by nickerick — COMMENTED** [review]
Beautiful. Huge fan of your javadocs. Even huger fan of your super detailed issue description.

File: server/src/main/java/com/skillmagnet/Course/Course.java (hunk 1 above)
Comment (sjdefran1) [reply server/src/main/java/com/skillmagnet/Course/Course.java]: Switched to `LocalDateTime` in the newest commit.

I couldn't find anything called just DateTIme but referenced this article [here](https://www.baeldung.com/java-8-date-time-intro)

LocalDatetime seems nice, lets you add/subtract minutes, hours, days, etc, which could come in handy. Section 7 of that article shows using the `DateTimeFormatter` to get custom formats.

File: server/src/main/java/com/skillmagnet/Course/CourseController.java (hunk 2 above)
Comment (sjdefran1) [reply server/src/main/java/com/skillmagnet/Course/CourseController.java]: Pretty sure I understand what you're saying. I went through and updated the URL's and am using `@PathVariable` now. Definitely makes sense and is more simplified.

File: server/src/main/java/com/skillmagnet/Course/CourseController.java (hunk 3 above)
Comment (sjdefran1) [reply server/src/main/java/com/skillmagnet/Course/CourseController.java:125]: Sounds good

File: server/src/main/java/com/skillmagnet/Course/Course.java [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+    private String description;
+    private String category;
+    private int views;
+    private int thumbsUp;
+    private int thumbsDown;
+
+    @CreationTimestamp
+    private LocalDateTime createdAt;
+    @CreationTimestamp
+    private LocalDateTime status;
```
Comment (sjdefran1) [inline server/src/main/java/com/skillmagnet/Course/Course.java:29]: LocalDateTime

File: server/src/main/java/com/skillmagnet/Course/CourseController.java [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+    @Autowired
+    CourseRepository courseRepository;
+
+    /**
+     * Retrieves specific course by id
+     * @param id - id of course to find
+     * @return - course object || NOT_FOUND
+     */
+    @GetMapping("/course/{id}")
+    public ResponseEntity<Course> getCourse(@PathVariable("id") int id){
```
Comment (sjdefran1) [inline server/src/main/java/com/skillmagnet/Course/CourseController.java:27]: Updated style of endpoint

**Review by nickerick — APPROVED** [review]
💪 perfect. good to merge whenever

**Pull request merged**